system/lib/wasmfs/backends/ignore_case_backend.cpp

```diff
@@ -20,105 +37,171 @@ std::string normalize(const std::string& name) {
+  return child;
+}
 
-  std::shared_ptr<Symlink> insertSymlink(const std::string& name,
-                                         const std::string& target) override {
-    auto name2 = normalize(name);
-    auto child = baseDirectory->locked().insertSymlink(name2, target);
-    if (child) {
-      insertChild(name2, child);
-      // Directory::Hanlde needs a parent
-      child->locked().setParent(cast<Directory>());
-    }
-    return child;
+std::shared_ptr<DataFile>
+IgnoreCaseDirectory::insertDataFile(const std::string& name, mode_t mode) {
+  auto normalized = normalize(name);
+  auto file = real->locked().insertDataFile(normalized, mode);
+  if (!file) {
+    return nullptr;
   }
+  file = virtualize(file, getBackend());
+  children[normalized] = {name, file};
+  return file;
+}
 
-  int insertMove(const std::string& name, std::shared_ptr<File> file) override {
-    auto newName = normalize(name);
-    // Remove entry with the new name (if any) from this directory.
-    if (auto err = removeChild(newName))
-      return err;
-    auto oldParent = file->locked().getParent()->locked();
-    auto oldName = normalize(oldParent.getName(file));
-    // Move in underlying directory.
-    if (auto err = baseDirectory->locked().insertMove(newName, file))
-      return err;
-    // Ensure old file was removed.
-    if (auto err = oldParent.removeChild(oldName))
-      return err;
-    // Cache file with the new name in this directory.
-    insertChild(newName, file);
-    file->locked().setParent(cast<Directory>());
-    return 0;
+std::shared_ptr<Directory>
+IgnoreCaseDirectory::insertDirectory(const std::string& name, mode_t mode) {
+  auto normalized = normalize(name);
+  auto dir = real->locked().insertDirectory(normalized, mode);
+  if (!dir) {
+    return nullptr;
   }
+  dir = virtualize(dir, getBackend());
+  children[normalized] = {name, dir};
+  return dir;
+}
 
-  int removeChild(const std::string& name) override {
-    auto name2 = normalize(name);
-    if (auto err = BaseClass::removeChild(name2))
-      return err;
-    return baseDirectory->locked().removeChild(name2);
+std::shared_ptr<Symlink>
+IgnoreCaseDirectory::insertSymlink(const std::string& name,
+                                   const std::string& target) {
+  auto normalized = normalize(name);
+  auto link = real->locked().insertSymlink(normalized, target);
+  if (!link) {
+    return nullptr;
   }
+  link = virtualize(link, getBackend());
+  children[normalized] = {name, link};
+  return link;
+}
 
-  ssize_t getNumEntries() override { return baseDirectory->locked().getNumEntries(); }
+int IgnoreCaseDirectory::insertMove(const std::string& name,
+                                    std::shared_ptr<File> file) {
+  auto normalized = normalize(name);
+  if (auto err = real->locked().insertMove(normalized, devirtualize(file))) {
+    return err;
+  }
+  auto oldParent =
+    std::static_pointer_cast<IgnoreCaseDirectory>(file->locked().getParent());
```

Comment:
Is it safe to use static cast here? Can we have a parent in different backend?

Reply:
Moves should always be between directories with the same backend, so this should be safe. Maintaining this invariant is why it's necessary to fully layer the virtual backend on top of the underlying backend and not mix them.

Reply:
Do moves assert on that property of it being in the same backend?

Reply:
Yes. More precisely, `__syscall_rename` returns `-EXDEV` if the parent backends do not match.